server/src-lib/Hasura/RQL/DDL/Metadata/Types.hs

```diff
@@ -0,0 +1,434 @@
+      <*> (o .:? "functions" >>= mapM (parseFunctions version))
+      <*> o .:? "remote_schemas"
+      <*> o .:? "query_collections"
+      <*> o .:? "allow_list"
+    where
+      parseFunctions = \case
+        MVVersion1 -> fmap FMVersion1 . parseJSON
+        MVVersion2 -> fmap FMVersion2 . parseJSON
+
+data ExportMetadata
+  = ExportMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''ExportMetadata)
+
+instance FromJSON ExportMetadata where
+  parseJSON _ = return ExportMetadata
+
+data ReloadMetadata
+  = ReloadMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''ReloadMetadata)
+
+instance FromJSON ReloadMetadata where
+  parseJSON _ = return ReloadMetadata
+
+data DumpInternalState
+  = DumpInternalState
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''DumpInternalState)
+
+instance FromJSON DumpInternalState where
+  parseJSON _ = return DumpInternalState
+
+data GetInconsistentMetadata
+  = GetInconsistentMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''GetInconsistentMetadata)
+
+instance FromJSON GetInconsistentMetadata where
+  parseJSON _ = return GetInconsistentMetadata
+
+data DropInconsistentMetadata
+ = DropInconsistentMetadata
+ deriving(Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''DropInconsistentMetadata)
+
+instance FromJSON DropInconsistentMetadata where
+  parseJSON _ = return DropInconsistentMetadata
+
+-- Functions enable encoding Metadata types and their dependants to ordered JSON
+
+replaceMetadataToOrdJSON :: ReplaceMetadata -> AO.Value
+replaceMetadataToOrdJSON ( ReplaceMetadata
+                               version
+                               tables
+                               functions
+                               remoteSchemas
+                               queryCollections
+                               allowlist
+                             ) = AO.object $
+                                 catMaybes [ versionPair
+                                           , tablesPair
+                                           , functionsPair
+                                           , remoteSchemasPair
+                                           , queryCollectionsPair
+                                           , allowlistPair
+                                           ]
+  where
+    versionPair = Just ("version", AO.toOrdered version)
+    tablesPair = Just ("tables", AO.array $ map tableMetaToOrdJSON tables)
+    functionsPair = functions >>= fmap ("functions",) . functionsMetadataToOrdJSON
+
+    remoteSchemasPair = remoteSchemas >>=
+                        listToMaybeOrdPair "remote_schemas" remoteSchemaQToOrdJSON
+
+    queryCollectionsPair = queryCollections >>=
+                           listToMaybeOrdPair "query_collections" createCollectionToOrdJSON
+
+    allowlistPair = allowlist >>= listToMaybeOrdPair "allowlist" AO.toOrdered
+
+tableMetaToOrdJSON :: TableMeta -> AO.Value
+tableMetaToOrdJSON ( TableMeta
```

Comment:
For all these little functions, I’d consider bumping them two indentation stops to be in the `where` clause of `replaceMetadataToOrdJSON`, just to emphasize that they’re really only to implement `replaceMetadataToOrdJSON`, as is mentioned in https://github.com/hasura/graphql-engine/pull/2175#issuecomment-514081273.